### ISSUE TITLE
Modules: Allow author to define separate typing for user-config and resulting runtime-config

### ISSUE THREAD

**Issue by BracketJohn** [nuxt/nuxt#20264]
### Goal

A module can specify separate typings for (currently inferred from moduleConfig) runtimeconfig types and for config that the user can provide.

### Description

When defining a module config like this:
```ts
interface ModuleOptions {
  baseURL?: string
}
```

and then setting a default value for base url in the `module.ts` when the developer did not specify anything, results in the runtime typing of still being `undefined | string`. It would be helpful to have a way to tell nuxt that at runtime the base url will definitely be `string`. For reasons on why this is helpful, look at the workarounds + counter-arguments listed below

Maybe there’s something I’m missing here / I hope that I can learn something I missed from this issue that allows to do this (: If so, it may be good to document for future module authors - I'd be willing to write that documentation.

Thanks!

### Workarounds

So far:
- add a helper that loads the runtime config and applies defaults (eg. via defu) at runtime
    - contra: increases bundle size, as it happens at runtime
    - contra: does not change runtime config typings globally, but only for consumers that use helper
- make all fields required
    - contra: Bad DX, for non-trivial examples (eg. the auth module) it becomes quite cumbersome for the dev to specify all arguments as soon as they want to specify a single one
-  keep fields optional, handle it in code
    - contra: bunch of `?.`s all over code -> bad for perf and maintainability

### Additional information

- [ ] Would you be willing to help implement this feature?
- [ ] Could this feature be implemented as a module?

### Final checks

- [X] Read the [contribution guide](https://nuxt.com/docs/community/contribution).
- [X] Check existing [discussions](https://github.com/nuxt/nuxt/discussions) and [issues](https://github.com/nuxt/nuxt/issues).

**Comment by BracketJohn** [comment]
I looked into it deeper, it turns out to rather be a problem how `defu` infers the types of deep discriminate unions. I cannot quite nail how / why this is happening - the following "fixes" it:
```ts
backend: defu(userOptions.backend, defaultsByBackend[selectedBackend]) as DeepRequired<AuthBackends>
```

"fixes" of course is n air-quotes because I do not think that using `as` actually qualifies as a solution, but rather a hack / workaround.

We can close this issue now, unless you see further action @danielroe

**Comment by danielroe** [comment]
@BracketJohn Basically do we need to fix this to convert a type from `string | undefined` -> `string` within `Defu` types when a subsequent type is guaranteed present? I think that should be doable within `defu`.

**Comment by danielroe** [comment]
I think this was implemented in https://github.com/nuxt/nuxt/pull/27520 (cc: @DamianGlowala). Let me know if not!

**Comment by DamianGlowala** [comment]
Indeed, using syntax from the linked PR should solve this issue. This should perhaps be mentioned somewhere in the module authors guide... ;)

**Comment by BracketJohn** [comment]
cc @phoenix-ru for potential impact/usefulness for nuxt-auth!

Thanks @danielroe & @DamianGlowala  for pushing this through (:
